x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java

```diff
@@ -102,6 +103,32 @@ public static AttributeSet references(List<? extends Expression> exps) {
         return set;
     }
 
+    public static AttributeSet referencesExcept(Set<? extends Expression> exps,
+                                                AttributeSet except) {
+        AttributeSet ret = new AttributeSet();
+        while(exps.size()>0){
```

Comment:
I don't get the logic here, why do you need the while loop and revisit the `exps` which have become `filteredExps`?

Reply:
It's a breadth-first search where some nodes are removed during the traversal. At each loop the exps represents the previous list of tree nodes and filteredExps the expressions that aren't in the except set. At the end of the loop exps is replaced with the children of each node. When a leaf is found, its references are added to ret.